File: messagebus/src/vespa/messagebus/network/rpcsendv1.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "rpcsendv1.h"
#include "rpcnetwork.h"
#include "rpcserviceaddress.h"
#include <vespa/messagebus/emptyreply.h>
#include <vespa/messagebus/error.h>
#include <vespa/vespalib/util/stringfmt.h>
#include <vespa/fnet/frt/reflection.h>

using vespalib::make_string;

namespace mbus {

namespace {

const char *METHOD_NAME   = "mbus.send1";
const char *METHOD_PARAMS = "sssbilsxi";
const char *METHOD_RETURN = "sdISSsxs";

}

bool RPCSendV1::isCompatible(vespalib::stringref method, vespalib::stringref request, vespalib::stringref respons)
{
    return  (method == METHOD_NAME) &&
            (request == METHOD_PARAMS) &&
            (respons == METHOD_RETURN);
}

const char *
RPCSendV1::getReturnSpec() const {
    return METHOD_RETURN;
}

void
RPCSendV1::build(FRT_ReflectionBuilder & builder)
{
    builder.DefineMethod(METHOD_NAME, METHOD_PARAMS, METHOD_RETURN, FRT_METHOD(RPCSendV1::invoke), this);
    builder.MethodDesc("Send a message bus request and get a reply back.");
    builder.ParamDesc("version", "The version of the message.");
    builder.ParamDesc("route", "Names of additional hops to visit.");
    builder.ParamDesc("session", "The local session that should receive this message.");
    builder.ParamDesc("retryEnabled", "Whether or not this message can be resent.");
    builder.ParamDesc("retry", "The number of times the sending of this message has been retried.");
    builder.ParamDesc("timeRemaining", "The number of milliseconds until timeout.");
    builder.ParamDesc("protocol", "The name of the protocol that knows how to decode this message.");
    builder.ParamDesc("payload", "The protocol specific message payload.");
    builder.ParamDesc("level", "The trace level of the message.");
    builder.ReturnDesc("version", "The lowest version the message was serialized as.");
    builder.ReturnDesc("retry", "The retry request of the reply.");
    builder.ReturnDesc("errorCodes", "The reply error codes.");
    builder.ReturnDesc("errorMessages", "The reply error messages.");
    builder.ReturnDesc("errorServices", "The reply error service names.");
    builder.ReturnDesc("protocol", "The name of the protocol that knows how to decode this reply.");
    builder.ReturnDesc("payload", "The protocol specific reply payload.");
    builder.ReturnDesc("trace", "A string representation of the trace.");
}

void
RPCSendV1::encodeRequest(FRT_RPCRequest &req, const vespalib::Version &version, const Route & route,
                         const RPCServiceAddress & address, const Message & msg, uint32_t traceLevel,
                         const PayLoadFiller &filler, duration timeRemaining) const
{

    FRT_Values &args = *req.GetParams();
    req.SetMethodName(METHOD_NAME);
    args.AddString(version.toString().c_str());
    args.AddString(route.toString().c_str());
    args.AddString(address.getSessionName().c_str());
    args.AddInt8(msg.getRetryEnabled() ? 1 : 0);
    args.AddInt32(msg.getRetry());
    args.AddInt64(vespalib::count_ms(timeRemaining));
    args.AddString(msg.getProtocol().c_str());
    filler.fill(args);
    args.AddInt32(traceLevel);
}

namespace {

class ParamsV1 : public RPCSend::Params
{
public:
    ParamsV1(const FRT_Values &args) : _args(args) { }

    uint32_t getTraceLevel() const override { return _args[8]._intval32; }
    bool useRetry() const override { return _args[3]._intval8 != 0; }
    uint32_t getRetries() const override { return _args[4]._intval32; }
    duration getRemainingTime() const override { return std::chrono::milliseconds(_args[5]._intval64); }

    vespalib::Version getVersion() const override {
        return vespalib::Version(vespalib::stringref(_args[0]._string._str, _args[0]._string._len));
    }
    vespalib::stringref getRoute() const override {
        return vespalib::stringref(_args[1]._string._str, _args[1]._string._len);
    }
    vespalib::stringref getSession() const override {
        return vespalib::stringref(_args[2]._string._str, _args[2]._string._len);
    }
    vespalib::stringref getProtocol() const override {
        return vespalib::stringref(_args[6]._string._str, _args[6]._string._len);
    }
    BlobRef getPayload() const override {
        return BlobRef(_args[7]._data._buf, _args[7]._data._len);
    }
private:
    const FRT_Values & _args;
};

}

std::unique_ptr<RPCSend::Params>
RPCSendV1::toParams(const FRT_Values &args) const
{
    return std::make_unique<ParamsV1>(args);
}


std::unique_ptr<Reply>
RPCSendV1::createReply(const FRT_Values & ret, const string & serviceName, Error & error, vespalib::Trace & trace) const
{
    vespalib::Version version          = vespalib::Version(ret[0]._string._str);
    double            retryDelay       = ret[1]._double;
    uint32_t         *errorCodes       = ret[2]._int32_array._pt;
    uint32_t          errorCodesLen    = ret[2]._int32_array._len;
    FRT_StringValue  *errorMessages    = ret[3]._string_array._pt;
    uint32_t          errorMessagesLen = ret[3]._string_array._len;
    FRT_StringValue  *errorServices    = ret[4]._string_array._pt;
    uint32_t          errorServicesLen = ret[4]._string_array._len;
    const char       *protocolName     = ret[5]._string._str;
    BlobRef payload(ret[6]._data._buf, ret[6]._data._len);
    const char       *traceStr         = ret[7]._string._str;

    Reply::UP reply;
    if (payload.size() > 0) {
        reply = decode(protocolName, version, payload, error);
    }
    if ( ! reply ) {
        reply = std::make_unique<EmptyReply>();
    }
    reply->setRetryDelay(retryDelay);
    for (uint32_t i = 0; i < errorCodesLen && i < errorMessagesLen && i < errorServicesLen; ++i) {
        reply->addError(Error(errorCodes[i], errorMessages[i]._str,
                              errorServices[i]._len > 0 ? errorServices[i]._str : serviceName.c_str()));
    }
    trace.addChild(TraceNode::decode(traceStr));
    return reply;
}

void
RPCSendV1::createResponse(FRT_Values & ret, const string & version, Reply & reply, Blob payload) const {
    ret.AddString(version.c_str());
    ret.AddDouble(reply.getRetryDelay());

    uint32_t         errorCount    = reply.getNumErrors();
    uint32_t        *errorCodes    = ret.AddInt32Array(errorCount);
    FRT_StringValue *errorMessages = ret.AddStringArray(errorCount);
    FRT_StringValue *errorServices = ret.AddStringArray(errorCount);
    for (uint32_t i = 0; i < errorCount; ++i) {
        errorCodes[i] = reply.getError(i).getCode();
        ret.SetString(errorMessages + i, reply.getError(i).getMessage().c_str());
        ret.SetString(errorServices + i, reply.getError(i).getService().c_str());
    }

    ret.AddString(reply.getProtocol().c_str());
    ret.AddData(std::move(payload.payload()), payload.size());
    if (reply.getTrace().getLevel() > 0) {
        ret.AddString(reply.getTrace().encode().c_str());
    } else {
        ret.AddString("");
    }
}

} // namespace mbus

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV1.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.messagebus.network.rpc;

import com.yahoo.component.Version;
import com.yahoo.jrt.DataValue;
import com.yahoo.jrt.DoubleValue;
import com.yahoo.jrt.Int32Array;
import com.yahoo.jrt.Int32Value;
import com.yahoo.jrt.Int64Value;
import com.yahoo.jrt.Int8Value;
import com.yahoo.jrt.Method;
import com.yahoo.jrt.Request;
import com.yahoo.jrt.StringArray;
import com.yahoo.jrt.StringValue;
import com.yahoo.jrt.Values;
import com.yahoo.messagebus.EmptyReply;
import com.yahoo.messagebus.Error;
import com.yahoo.messagebus.Message;
import com.yahoo.messagebus.Reply;
import com.yahoo.messagebus.Trace;
import com.yahoo.messagebus.TraceNode;
import com.yahoo.messagebus.routing.Route;
import com.yahoo.text.Utf8Array;

/**
 * Implements the request adapter for method "mbus.send1".
 *
 * @author Simon Thoresen Hult
 */
public class RPCSendV1 extends RPCSend {

    private final String METHOD_NAME = "mbus.send1";
    private final String METHOD_PARAMS = "sssbilsxi";
    private final String METHOD_RETURN = "sdISSsxs";

    @Override
    protected String getReturnSpec() { return METHOD_RETURN; }
    @Override
    protected Method buildMethod() {

        Method method = new Method(METHOD_NAME, METHOD_PARAMS, METHOD_RETURN, this);
        method.methodDesc("Send a message bus request and get a reply back.");
        method.paramDesc(0, "version", "The version of the message.")
                .paramDesc(1, "route", "Names of additional hops to visit.")
                .paramDesc(2, "session", "The local session that should receive this message.")
                .paramDesc(3, "retryEnabled", "Whether or not this message can be resent.")
                .paramDesc(4, "retry", "The number of times the sending of this message has been retried.")
                .paramDesc(5, "timeRemaining", "The number of milliseconds until timeout.")
                .paramDesc(6, "protocol", "The name of the protocol that knows how to decode this message.")
                .paramDesc(7, "payload", "The protocol specific message payload.")
                .paramDesc(8, "level", "The trace level of the message.");
        method.returnDesc(0, "version", "The lowest version the message was serialized as.")
                .returnDesc(1, "retryDelay", "The retry request of the reply.")
                .returnDesc(2, "errorCodes", "The reply error codes.")
                .returnDesc(3, "errorMessages", "The reply error messages.")
                .returnDesc(4, "errorServices", "The reply error service names.")
                .returnDesc(5, "protocol", "The name of the protocol that knows how to decode this reply.")
                .returnDesc(6, "payload", "The protocol specific reply payload.")
                .returnDesc(7, "trace", "A string representation of the trace.");
        return method;
    }
    @Override
    protected Request encodeRequest(Version version, Route route, RPCServiceAddress address, Message msg,
                                    long timeRemaining, byte[] payload, int traceLevel) {
        Request req = new Request(METHOD_NAME);
        Values v = req.parameters();
        v.add(new StringValue(version.toUtf8()));
        v.add(new StringValue(route.toString()));
        v.add(new StringValue(address.getSessionName()));
        v.add(new Int8Value(msg.getRetryEnabled() ? (byte)1 : (byte)0));
        v.add(new Int32Value(msg.getRetry()));
        v.add(new Int64Value(timeRemaining));
        v.add(new StringValue(msg.getProtocol()));
        v.add(new DataValue(payload));
        v.add(new Int32Value(traceLevel));
        return req;
    }

    @Override
    protected Reply createReply(Values ret, String serviceName, Trace trace) {
        Version version = new Version(ret.get(0).asUtf8Array());
        double retryDelay = ret.get(1).asDouble();
        int[] errorCodes = ret.get(2).asInt32Array();
        String[] errorMessages = ret.get(3).asStringArray();
        String[] errorServices = ret.get(4).asStringArray();
        Utf8Array protocolName = ret.get(5).asUtf8Array();
        byte[] payload = ret.get(6).asData();
        String replyTrace = ret.get(7).asString();

        // Make sure that the owner understands the protocol.
        Reply reply = null;
        Error error = null;
        if (payload.length > 0) {
            Object retval = decode(protocolName, version, payload);
            if (retval instanceof Reply) {
                reply = (Reply) retval;
            } else {
                error = (Error) retval;
            }
        }
        if (reply == null) {
            reply = new EmptyReply();
        }
        if (error != null) {
            reply.addError(error);
        }
        reply.setRetryDelay(retryDelay);
        for (int i = 0; i < errorCodes.length && i < errorMessages.length; i++) {
            reply.addError(new Error(errorCodes[i], errorMessages[i],
                    errorServices[i].length() > 0 ? errorServices[i] : serviceName));
        }
        if (trace.getLevel() > 0) {
            trace.getRoot().addChild(TraceNode.decode(replyTrace));
        }
        return reply;
    }

    protected Params toParams(Values args) {
        Params p = new Params();
        p.version = new Version(args.get(0).asUtf8Array());
        p.route = args.get(1).asString();
        p.session = args.get(2).asString();
        p.retryEnabled = (args.get(3).asInt8() != 0);
        p.retry = args.get(4).asInt32();
        p.timeRemaining = args.get(5).asInt64();
        p.protocolName = args.get(6).asUtf8Array();
        p.payload = args.get(7).asData();
        p.traceLevel = args.get(8).asInt32();
        return p;
    }

    @Override
    protected void createResponse(Values ret, Reply reply, Version version, byte [] payload) {
        int[] eCodes = new int[reply.getNumErrors()];
        String[] eMessages = new String[reply.getNumErrors()];
        String[] eServices = new String[reply.getNumErrors()];
        for (int i = 0; i < reply.getNumErrors(); ++i) {
            Error error = reply.getError(i);
            eCodes[i] = error.getCode();
            eMessages[i] = error.getMessage();
            eServices[i] = error.getService() != null ? error.getService() : "";
        }
        ret.add(new StringValue(version.toUtf8()));
        ret.add(new DoubleValue(reply.getRetryDelay()));
        ret.add(new Int32Array(eCodes));
        ret.add(new StringArray(eMessages));
        ret.add(new StringArray(eServices));
        ret.add(new StringValue(reply.getProtocol()));
        ret.add(new DataValue(payload));
        ret.add(new StringValue(reply.getTrace().getRoot() != null ? reply.getTrace().getRoot().encode() : ""));
    }

}
